src/packet_analysis/protocol/ip/IPBasedAnalyzer.h

```diff
@@ -0,0 +1,133 @@
+
+namespace zeek::packet_analysis::IP {
+
+/**
+ * A base class for any packet analyzer based on IP. This is used by default by
+ * the TCP, UDP, and ICMP analyzers to reduce a large amount of duplicated code
+ * that those plugins have in common.
+ */
+class IPBasedAnalyzer : public Analyzer {
+public:
+	~IPBasedAnalyzer() override;
+
+	bool AnalyzePacket(size_t len, const uint8_t* data, Packet* packet) override;
+
+	/**
+	 * Returns true if the analyzer determines that in fact a new
+	 * connection has started without the connection statement having
+	 * terminated the previous one, i.e., the new data is arriving at
+	 * what's the analyzer for the previous instance. This is used only
+	 * for TCP.
+	 */
+	virtual bool IsReuse(double t, const u_char* pkt) { return false; }
+
+protected:
+
+	IPBasedAnalyzer(const char* name, TransportProto proto, uint32_t mask);
+
+	/**
+	 * Parse the header from the packet into a ConnTuple object.
+	 */
+	virtual bool BuildConnTuple(size_t len, const uint8_t* data, Packet* packet,
+	                            ConnTuple& tuple) = 0;
+
+	/**
+	 * Continues process of packet after the connection has been inserted into the
+	 * session manager. This should be implemented by all child classes.
+	 *
+	 * @param conn The connection currently being processed.
+	 * @param t The timestamp for the current packet.
+	 * @param is_orig Flag denoting whether this packet is from the originator of
+	 * the connection.
+	 * @param remaining The remaining about of data in the packet.
+	 * @param pkt The packet being processed.
+	 */
+	virtual void DeliverPacket(Connection* conn, double t, bool is_orig, int remaining,
+	                           Packet* pkt) {}
+
+	/**
+	 * Upon seeing the first packet of a connection, checks whether we want
+	 * to analyze it (e.g. we may not want to look at partial connections)
+	 * and, if yes, whether we should flip the roles of originator and
+	 * responder based on known ports and such.
+	 *
+	 * @param src_port The source port of the connection.
+	 * @param dst_port The destination port of the connection.
+	 * @param data The payload data for the packet being processed.
+	 * @param flip_roles Return value if the roles should be flipped.
+	 * @return True if the connection is wanted. False otherwise.
+	 */
+	virtual bool WantConnection(uint16_t src_port, uint16_t dst_port,
+	                            const u_char* data, bool& flip_roles) const
+		{
+		flip_roles = false;
+		return true;
+		}
+
+	/**
+	 * Returns an analyzer adapter appropriate for this IP-based analyzer. This adapter
+	 * is used to hook into the session analyzer framework. This function can also be used
+	 * to do any extra initialization of connection timers, etc.
+	 */
+	virtual AnalyzerAdapter* MakeAnalyzerAdapter(Connection* conn) { return nullptr; }
+
+	/**
+	 * Returns a PIA appropriate for this IP-based analyzer.
+	 */
+	virtual analyzer::pia::PIA* MakePIA(Connection* conn) { return nullptr; }
+
```

Comment:
Add a comment to these two saying whether derived class must implement these and return instances. `MakeAnalyzerAdapter` probably needs to be, but we should be ok without a PIA?

Reply:
Oh, and if one of the two *must* be implemented make it pure. Maybe the absence of that means even the `MakeAnalyzerAdapter` is optional?

Reply:
`MakeAnalyzerAdapter` can be made pure virtual, but not until the TCP work is done. I'll add a comment to note that it needs to be fixed before that work is done.